keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1344 @@
+time and computational resources wasted due to unnecessary retries of containers
+destined to fail due to software bugs.
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+Running a large computational workload, comprising thousands of pods on
+thousands of nodes requires usage of pod restart policies in order
+to account for infrastructure failures.
+
+Currently, kubernetes Job API offers a way to account for infrastructure
+failures by setting `.backoffLimit > 0`. However, this mechanism intructs the
+job controller to restart all failed pods - regardless of the root cause
+of the failures. Thus, in some scenarios this leads to unnecessary
+restarts of many pods, resulting in a waste of time and computational
+resources. What makes the restarts more expensive is the fact that the
+failures may be encountered late in the execution time of a program.
+
+Sometimes it can be determined from containers exit codes
+that the root cause of a failure is in the executable and the
+job is destined to fail regardless of the number of retries. However, since
+the large workloads are often scheduled to run over night or over the
+weekend, there is no human assistance to terminate such a job early.
+
+The need for solving the problem has been emphasized by the kubernetes
+community in the issues, see: [#17244](https://github.com/kubernetes/kubernetes/issues/17244) and [#31147](https://github.com/kubernetes/kubernetes/issues/31147).
+
+Some third-party frameworks have implemented retry policies for their pods:
+- [TensorFlow Training (TFJob)](https://www.kubeflow.org/docs/components/training/tftraining/)
+- [Argo workflow](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml) ([example](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml))
+
+Additionally, some pod failures are not linked with the container execution,
+but rather with the internal kubernetes cluster management (see:
+[Scheduling, Preemption and Eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/)).
```

Comment:
Beyond the errors around Scheduling, Pre-emption and Eviction, there could be numerous other failures not related to container execution that would lead to repeated failures that would require human intervention to recover from. Some examples:
1. Incorrect image specs on containers
2. Incorrect secret/configmap references
3. Incorrect PVC references.
4. Inability of a CSI plugin to attach/mount necessary inline/persistent volumes 
5. General admission failure due to violation of configured policies through OPA/other webhooks.

Are errors like above in scope of the kind of failures that need to be distinguished by the job controller?

Reply:
In cases 1-3 (which I tested) pods get stuck in the `status.phase=Pending` rather than `Failed`. In case 1 we get a nice dedicated `.status.containerStatuses.waiting.reason=ErrImagePull`, but in cases 2-3 pods get stuck in the `CreatingContainer` status. 

Handling of such errors is out of scope for this KEP. Such configuration issues should be first recognized as Pod failures (not pending pods) which seems deserving a KEP on its own.